cpp/src/arrow/json/parser.h

```diff
@@ -0,0 +1,142 @@
+#include <unordered_map>
+
+#include "arrow/builder.h"
+#include "arrow/json/options.h"
+#include "arrow/status.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class MemoryPool;
+class RecordBatch;
+
+namespace json {
+
+/// \class AdaptiveArrayBuilder
+/// \brief Base class for array builders in contexts where the
+///        final type is unknown
+class AdaptiveArrayBuilder {
+ public:
+  AdaptiveArrayBuilder(std::shared_ptr<DataType> type) : type_(type) {}
+
+  virtual ~AdaptiveArrayBuilder() = default;
+
+  /// Complete the built array
+  virtual Status Finish(std::shared_ptr<Array>* out) = 0;
+
+  /// If necessary, promote this builder to accommodate the given type.
+  /// It is not guaranteed that type() == type after this method is called.
+  /// It is possible that a new builder must be constructed
+  /// in order to accomodate this fallback. If this occurs,
+  /// *out will point to the new builder and the object against
+  /// which this method was called should be considered moved from.
+  virtual Status MaybePromoteTo(std::shared_ptr<DataType> type,
+                                std::unique_ptr<AdaptiveArrayBuilder>* out) = 0;
+
+  /// The current type of this builder
+  ///
+  /// \warning this property may be lazily updated, it is only guaranteed
+  /// to be accurate after a call to UpdateType()
+  std::shared_ptr<DataType> type() { return type_; }
+
+  int64_t length() { return length_; }
+
+  /// Force update of built type. If an implementation of AdaptiveArrayBuilder updates
+  /// type lazily, UpdateType() must be overridden. Implementations may not require that
+  /// it be called before Finish() or MaybePromoteTo()
+  virtual Status UpdateType() { return Status::OK(); }
+
+ protected:
+  ARROW_DISALLOW_COPY_AND_ASSIGN(AdaptiveArrayBuilder);
+
+  int64_t length_ = 0;
+  std::shared_ptr<DataType> type_;
+};
+
+constexpr int32_t kMaxParserNumRows = 100000;
+
+/// \class BlockParser
+/// \brief A reusable block-based parser for JSON data
+///
+/// The parser takes a block of newline delimited JSON data and extracts
+/// keys and value pairs, inserting into provided ArrayBuilders.
+/// Parsed data is own by the
+/// parser, so the original buffer can be discarded after Parse() returns.
+class ARROW_EXPORT BlockParser {
+ public:
+  explicit BlockParser(ParseOptions options, int32_t num_cols = -1,
+                       int32_t max_num_rows = kMaxParserNumRows);
+  explicit BlockParser(MemoryPool* pool, ParseOptions options, int32_t num_cols = -1,
+                       int32_t max_num_rows = kMaxParserNumRows);
+
+  /// \brief Parse a block of data
+  ///
+  /// Parse a block of JSON data, ingesting up to max_num_rows rows.
+  /// The number of bytes actually parsed is returned in out_size.
+  Status Parse(const char* data, uint32_t size, uint32_t* out_size);
+
+  /// \brief Extract parsed data as a RecordBatch
+  Status Finish(std::shared_ptr<RecordBatch>* parsed) {
```

Comment:
Type inference will have to work at the level of all blocks (which there are several of, otherwise you can't parse in parallel). So you can't return a `RecordBatch` here, since you can't infer the column types from a single block.

Of course, this assumes we want to do type inference at all, since JSON is already (partially) typed. @wesm what is the plan here?

Reply:
sorry, nevermind. I misread your comment

Reply:
Passing a schema is just one way to indicate field types -- also if you _do_ pass a schema, then fields not contained in the schema should be discarded. I think we also want to allow for passing types for one or more fields (allowing others to be inferred) in the same style as the CSV reader.

To give an example of how this might be useful: if we have dates in a file represented in some particular way (e.g. something we can parse with strptime) then we could indicate the conversion rule for just that field

This doesn't all have to get done in a single PR, but it's worth keeping in mind for structuring the abstractions so that this functionality can be developed incrementally.

Reply:
Could you add this to the design Google doc?

Reply:
yes, will do

Reply:
https://avro.apache.org/docs/1.8.1/spec.html#Schema+Resolution